presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -112,10 +134,39 @@ public void process(Iterable<QueryExecution> queries)
             if (bytes > queryMemoryLimit) {
                 DataSize maxMemory = succinctDataSize(queryMemoryLimit, BYTE);
                 query.fail(exceededGlobalLimit(maxMemory));
+                queryKilled = true;
             }
         }
         clusterMemoryUsageBytes.set(totalBytes);
 
+        boolean shouldKillQuery = nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) == 1 && isClusterOutOfMemory();
+        boolean lastKilledQueryIsGone = lastKilledQuery == null;
+
+        if (lastKilledQuery != null) {
```

Comment:
Actually no. This modifies `lastKilledQueryIsGone` based on whether that query has terminated on all workers